cmd/contour/servecontext_test.go

```diff
@@ -100,3 +104,98 @@ func TestServeContextTLSParams(t *testing.T) {
 		})
 	}
 }
+
+func TestConfigFileDefaultOverrideImport(t *testing.T) {
+	tests := map[string]struct {
+		yamlIn []byte
+		want   func() *serveContext
+	}{
+		"empty configuration": {
+			yamlIn: []byte(``),
+			want:   newServeContext,
+		},
+		"defaults in yaml": {
+			yamlIn: []byte(`
+incluster: false
+disablePermitInsecure: false
+leaderelection:
+  configmap-name: contour
+  configmap-namespace: heptio-contour
+  lease-duration: 15s
+  renew-deadline: 10s
+  retry-period: 2s
+`),
+			want: newServeContext,
+		},
+		"blank tls configuration": {
+			yamlIn: []byte(`
+tls:
+`),
+			want: newServeContext,
+		},
+		"tls configuration only": {
+			yamlIn: []byte(`
+tls:
+  minimum-protocol-version: 1.2
+`),
+			want: func() *serveContext {
+				ctx := newServeContext()
+				ctx.TLSConfig = TLSConfig{
+					MinimumProtocolVersion: "1.2",
+				}
+				return ctx
+			},
+		},
+		"leader election namespace and configmap only": {
+			yamlIn: []byte(`
+leaderelection:
+  configmap-name: foo
+  configmap-namespace: bar
+`),
+			want: func() *serveContext {
+				ctx := newServeContext()
+				ctx.LeaderElectionConfig.Name = "foo"
+				ctx.LeaderElectionConfig.Namespace = "bar"
+				return ctx
+			},
+		},
+		"leader election all fields set": {
+			yamlIn: []byte(`
+leaderelection:
+  configmap-name: foo
+  configmap-namespace: bar
+  lease-duration: 600s
+  renew-deadline: 500s
+  retry-period: 60s
+`),
+			want: func() *serveContext {
+				ctx := newServeContext()
+				ctx.LeaderElectionConfig.Name = "foo"
+				ctx.LeaderElectionConfig.Namespace = "bar"
+				ctx.LeaderElectionConfig.LeaseDuration = 600 * time.Second
+				ctx.LeaderElectionConfig.RenewDeadline = 500 * time.Second
+				ctx.LeaderElectionConfig.RetryPeriod = 60 * time.Second
+				return ctx
+			},
+		},
+	}
+	for name, tc := range tests {
+		t.Run(name, func(t *testing.T) {
+			got := newServeContext()
+			err := yaml.Unmarshal(tc.yamlIn, got)
+			checkErr(t, err)
+			want := tc.want()
+
+			if diff := cmp.Diff(*want, *got, cmp.AllowUnexported(serveContext{})); diff != "" {
```

Comment:
Maybe we should just export ServeContext. It's in `package main` so nobody can import it anyway. If we export it then the comments will show up in godoc which is probably a good thing.

Reply:
AllowUnexported is actually talking about allowing comparison of the unexported fields of `serveContext` here.
Not sure about changing all of them to be exported just for this test case.

Reply:
true, ok, skip this and lets table the discussion of the value of go-cmp/cmp until after 1.0.